src/coreclr/jit/codegenxarch.cpp

```diff
@@ -7532,47 +7532,53 @@ void CodeGen::genIntToFloatCast(GenTree* treeNode)
     // integral to floating-point conversions all have RMW semantics if VEX support
     // is not available
 
-    bool isRMW = !compiler->canUseVexEncoding();
-    inst_RV_RV_TT(ins, emitTypeSize(srcType), targetReg, targetReg, op1, isRMW, INS_OPTS_NONE);
+    const bool isRMW = !compiler->canUseVexEncoding();
 
     // Handle the case of srcType = TYP_ULONG. SSE2 conversion instruction
     // will interpret ULONG value as LONG.  Hence we need to adjust the
     // result if sign-bit of srcType is set.
     if (srcType == TYP_ULONG)
     {
-        // The instruction sequence below is less accurate than what clang
-        // and gcc generate. However, we keep the current sequence for backward compatibility.
-        // If we change the instructions below, FloatingPointUtils::convertUInt64ToDobule
-        // should be also updated for consistent conversion result.
         assert(dstType == TYP_DOUBLE);
         assert(op1->isUsedFromReg());
 
-        // Set the flags without modifying op1.
-        // test op1Reg, op1Reg
-        inst_RV_RV(INS_test, op1->GetRegNum(), op1->GetRegNum(), srcType);
+        // The following LONG->DOUBLE cast machinery is based on clang's implementation
```

Comment:
I wonder if doing this in codegen is better than doing it in lowering or an inlinable managed implementation instead.

For example, this hinders LSRAs ability to optimize (such as if `argReg` was last use since temp registers are always delay free) and will also impact the ability to easily pick the right instructions for each node, which will be more relevant when APX comes online.

The logic here isn't really that complex and basically boils down to:
```csharp
static double ConvertToDouble(ulong value)
{
    long signedValue = (long)value;

    if (signedValue >= 0)
    {
        return signedValue;
    }
    signedValue = (signedValue >>> 1) | (signedValue & 1);

    double tmp = signedValue;
    return tmp + tmp;
}
```

The JIT generates pretty ideal code for this already (this is for SSE2 baseline):
```asm
; Method Program:ConvertToDouble(ulong):double (FullOpts)
G_M000_IG01:                ;; offset=0x0000

G_M000_IG02:                ;; offset=0x0000
       test     rcx, rcx
       jl       SHORT G_M000_IG05

G_M000_IG03:                ;; offset=0x0005
       xorps    xmm0, xmm0
       cvtsi2sd  xmm0, rcx

G_M000_IG04:                ;; offset=0x000D
       ret      

G_M000_IG05:                ;; offset=0x000E
       mov      rax, rcx
       shr      rax, 1
       and      rcx, 1
       or       rax, rcx
       xorps    xmm0, xmm0
       cvtsi2sd  xmm0, rax
       addsd    xmm0, xmm0

G_M000_IG06:                ;; offset=0x0027
       ret      
; Total bytes of code: 40
```

The only real pessimization is that `cvtsi2sd` can avoid the `xorps` and leave the upper bits undefined instead to save a few extra bytes of codegen. Given we already have a branch for `jns` its also not clear that using `cmov` is preferable over doing the early out instead and minimally duplicating the logic 

-------------

Notably we can get the JIT to generate `cmov` instead via:
```csharp
static double ConvertToDouble2(ulong value)
{
    long signedValue = (long)value;

    if (signedValue < 0)
    {
        signedValue = (signedValue >>> 1) | (signedValue & 1);
    }

    double result = signedValue;

    if (signedValue < 0)
    {
        result += result;
    }

    return result;
}
```

which generates
```asm
; Method Program:ConvertToDouble2(ulong):double (FullOpts)
G_M000_IG01:                ;; offset=0x0000

G_M000_IG02:                ;; offset=0x0000
       mov      rax, rcx
       mov      rdx, rcx
       shr      rdx, 1
       and      rcx, 1
       or       rcx, rdx
       test     rax, rax
       cmovl    rax, rcx
       xorps    xmm0, xmm0
       cvtsi2sd  xmm0, rax
       test     rax, rax
       jge      SHORT G_M000_IG04

G_M000_IG03:                ;; offset=0x0024
       addsd    xmm0, xmm0

G_M000_IG04:                ;; offset=0x0028
       ret      
; Total bytes of code: 41
```

The same `xorps` bit exists as does us missing we can reuse the flags from the previous test, since no intermediate operations overwrote them. But both are general purpose opts we can make to then get identical codegen to clang. -- We also pick the `l` and `ge` conditions, rather than the `ns` condition, but that's just because we don't model that `test` always clears `OF` and so we can just check `SF` rather than `SF != OF`

Reply:
I agree that Lower sounds like a better place, the main problem that we can't introduce control flow (basic blocks) there so I don't know how to do that jump, perhaps can be made branchless with some efforts with SIMD, but I guess we talk about a relatively rare case of `ulong -> double/float` cast on X64 without AVX512, so not sure it's worth the efforts. Since this change is more about correctness I propose we think about a more performant (in case if the current approach is indeed slow) solution separately

Reply:
> If we could just do the managed implementation and the JIT could inline that managed helper,

We currently have a single GT_CAST node that is easy to hoist/cse etc. An inlined control flow from a managed helper will be, sadly, hard to work with 🙁